integration/cluster.go

```diff
@@ -0,0 +1,273 @@
+// AcceptancePolicy is set to most permissive mode on first manager node added.
+func NewCluster() *Cluster {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &Cluster{
+		ctx:    ctx,
+		cancel: cancel,
+		nodes:  make(map[string]*TestNode),
+		errs:   make(chan error, 1024),
+	}
+	c.api = &dummyAPI{c: c}
+	return c
+}
+
+// Stop makes best effort to stop all nodes and close connections to them.
+func (c *Cluster) Stop() error {
+	c.cancel()
+	for _, n := range c.nodes {
+		if err := n.Stop(); err != nil {
+			return err
+		}
+	}
+	c.wg.Wait()
+	close(c.errs)
+	for err := range c.errs {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RandomManager choses random manager from cluster.
+func (c *Cluster) RandomManager() *TestNode {
+	var managers []*TestNode
+	for _, n := range c.nodes {
+		if n.config.IsManager {
+			managers = append(managers, n)
+		}
+	}
+	idx := rand.Intn(len(managers))
+	return managers[idx]
+}
+
+// AddManager adds node with Manager role(both agent and manager).
+func (c *Cluster) AddManager() error {
+	// first node
+	var n *TestNode
+	if len(c.nodes) == 0 {
+		node, err := newNode(true, "")
+		if err != nil {
+			return err
+		}
+		n = node
+	} else {
+		joinAddr, err := c.RandomManager().node.RemoteAPIAddr()
+		if err != nil {
+			return err
+		}
+		node, err := newNode(true, joinAddr)
+		if err != nil {
+			return err
+		}
+		n = node
+	}
+	c.wg.Add(1)
+	go func() {
+		c.errs <- n.node.Start(c.ctx)
+		c.wg.Done()
+	}()
+
+	select {
+	case <-n.node.Ready():
+	case <-time.After(opsTimeout):
+		return fmt.Errorf("node did not ready in time")
+	}
+
+	// change acceptance policy on cluster creation
+	if len(c.nodes) == 0 {
+		// we retry this, because sequence number can change
+		var ok bool
+		for retry := 0; retry < 5; retry++ {
+			if err := n.SetAcceptancePolicy(); err != nil {
+				if grpc.ErrorDesc(err) == "update out of sequence" {
+					continue
+				}
+				return fmt.Errorf("set acceptance policy: %v", err)
+			}
+			ok = true
+			break
+		}
+		if !ok {
+			return fmt.Errorf("set acceptance policy, got sequence error 5 times")
+		}
+	}
+	c.nodes[n.node.NodeID()] = n
+	return nil
+}
+
+// AddAgent adds node with Agent role(doesn't participate in raft cluster).
+func (c *Cluster) AddAgent() error {
+	// first node
+	var n *TestNode
+	if len(c.nodes) == 0 {
+		return fmt.Errorf("there is no manager nodes")
+	}
+	joinAddr, err := c.RandomManager().node.RemoteAPIAddr()
+	if err != nil {
+		return err
+	}
+	node, err := newNode(false, joinAddr)
+	if err != nil {
+		return err
+	}
+	n = node
+	c.wg.Add(1)
+	go func() {
+		c.errs <- n.node.Start(c.ctx)
+		c.wg.Done()
+	}()
+
+	select {
+	case <-n.node.Ready():
+	case <-time.After(opsTimeout):
+		return fmt.Errorf("node is not ready in time")
+	}
+	c.nodes[n.node.NodeID()] = n
+	return nil
+}
+
+// CreateService creates dummy service.
+func (c *Cluster) CreateService(name string, instances int) (string, error) {
```

Comment:
Currently there are `replicated` and `global` services. There may be more to come.